### PR TITLE
fix(dev): CTL-2068 — a claim the pipeline can overwrite is not a claim

### PR DESCRIPTION
**Pull request by ryanrozich**
## The collision, from the replica

Twice in two nights (CTC-776, CTC-787) a fleet worker built a ticket a human-facing lane was
already building — after the lane claimed it exactly as the rule says. `issue_history` has the
second one, and it is the whole mechanism:

| when | actor | move | |
| -- | -- | -- | -- |
| 1787132205283 | **Ryan Rozich** | `Validate → Implement` | the lane claims it |
| 1787132279778 | **Catalyst Cloud** | `Implement → Research` | ⛔ 74 seconds later, the pipeline drags it BACKWARDS |
| … | | | a worker dispatches `research` and opens the duplicate #993 |

`CTL-758` already refuses a backward Linear write — but **only out of a TERMINAL state**
(`Done`/`Canceled`), so `Implement → Research` sails straight through it. This generalizes that
guard along the one axis that makes it safe to widen.

## ⛔ Why a blanket "never regress" is wrong, and what the guard keys on instead

The pipeline **legitimately** re-runs earlier phases — the L3 destroy-and-recreate on
`research`/`plan`, the `verify ⇄ remediate` cycle — and those writes *are* backward moves. A rule
that refused every regression would break recovery.

What separates the two is not the direction of the move but **who last moved the ticket**. Lanes
write state through `linearis` with a human user id; the daemon writes as the app-actor whose id
is the already-configured `botUserId`. On CTC-787 those are `c2a8cc92…` (Ryan Rozich) and
`78f8f491…` (Catalyst Cloud) — read off the replica, not assumed. The guard takes that Set as
input and hardcodes nothing.

## Two enforcement points, because refusing the write does not stop the work

⭐ On CTC-787 the fleet held the ticket **in its pipeline the whole time**. Refusing the state
write alone would have made the collision merely *visible* — the worker would still have run and
still opened the duplicate PR. So the same question is asked at both sites:

| site | question |
| -- | -- |
| **write guard** (`linear-write.mjs`) | may the pipeline write this phase's state over the current one? |
| **dispatch veto** (`scheduler.mjs` → `dispatchAndVerify`) | may the pipeline run this phase at all? |

*If the pipeline may not move a ticket back to a phase's state, it has no business running that
phase.* One rule, two enforcement points.

The write guard is applied on **both rungs CTL-758 uses**, for the same documented reason: the
pre-read is fail-open and returns `null` when it cannot answer, so `--resolve-only`'s fresher state
gets its own re-application. The first rung also means the guard holds on **every transport**
(shadow proxy, proxy off, plain shell) — a guard wired only to the enforce path would be live on
both minis today purely because that is how they happen to be configured, which is the
"held by an accident" shape this ticket is about.

## Fails OPEN, and every decline is NAMED

This is a refusal guard on the pipeline's own writes and dispatches, so a guard that cannot answer
must not block. Every such case returns `INCONCLUSIVE` **with a reason** rather than a quiet allow,
so "I could not look" is never recorded as "nothing is wrong":

- no history · no bot ids · an unranked state · an unreadable/throwing replica
- **`triage`** — writes no status, and is where fleet work legitimately begins
- ⛔ **an EMPTY `botUserIds` set** — with no known fleet ids *every* actor looks like a lane, and the
  guard would refuse the pipeline's every legitimate backward move. That is the dangerous
  direction, so it is inconclusive, never a licence to refuse.
- ⭐ **`Todo → Research`** — `Todo` is deliberately unranked, so a human queueing work is not read as
  a claim and the fleet can still start it.

## Ordering, derived rather than declared

`buildStateRank` (current side, by state NAME) and `buildKeyRank` (target side, by linear KEY) are
both derived from `PHASE_LINEAR_KEY` at the **MIN** phase index. Four phases write `inReview`, so
`PR` ranks at 6, not teardown's 9 — a max/last-wins rank would make a legitimate
`monitor-merge → pr` write look like a regression.

Only the states the pipeline actually writes are ranked. `Todo`/`Backlog`/`Triage`/`Done` are
absent on purpose, and an absent state abstains instead of being ordered at some invented position.
It reads **one** rung of the state-name chain, never the four-rung precedence chain
`linear-transition.sh` owns — so this guard can only ever abstain where that chain would have
disagreed, never contradict it.

## Armed by default

Installed unconditionally, not shadow-first. The refusal is narrow and every unjudgeable case
allows, so there is no posture for it to be wrong in — while a shadow default would reproduce this
repo's recurring failure of shipping a correct mechanism nobody ever arms. The **dispatch** half
withholds *work* rather than just a status write, so it keeps an operator kill switch,
`CATALYST_LANE_CLAIM_DISPATCH_GUARD=off`, which leaves the write guard armed.

`botUserIds` is passed **by reference** — the stable Set `refreshBotIdentities` fills in place on
every cluster-sync tick — so an app-actor re-mint keeps the guard current instead of freezing it at
boot. A copy would make the guard read every post-rotation fleet write as a lane's. There is a test
for exactly that.

## Two things found on the way

1. ⛔ **`applyPhaseStatus` accepted six named options and passed five on**, silently dropping
   `fetchState`. Invisible in production — the default *is* the production value — and fatal to any
   test that injected it: the injection appeared to work while actually measuring the default path.
   Caught by asserting **the shell is never spawned**, not by reading the returned shape.
2. ⛔ **`scheduler.test.mjs`, where `dispatchAndVerify`'s other tests live, is EXCLUDED from
   `execution-core-tests.yml`** (flaky `fs.watch` timers) — so a veto test placed there would never
   run in CI. Both new test files carry their own harness and are **added to the workflow's stable
   list in this PR**. (A guard nobody runs is a check that cannot fail — the same shape as the
   defect being fixed.)

## Verification

**42 tests. Nine mutations red**, each with a discriminating count: the regression comparator
(`<` → `<=`), the fleet-actor branch inverted, MIN→MAX ranking, the empty-bot-ids guard removed,
unranked-current→refuse, **both** write rungs, the dispatch veto, and the option forwarding.

Controls throughout, because the interesting risk is a guard that refuses everything:

- the **identical** regression authored by the FLEET is ALLOWED
- a **forward** move under a live lane claim is ALLOWED
- `Todo → Research` is ALLOWED
- with **no guard installed** the same tick dispatches (pre-CTL-2068 behaviour)
- the kill switch disables **only** the dispatch veto; the write guard stays armed

Full CI stable suite: **8194 pass / 2 fail**. Both failures are **pre-existing and unrelated**,
verified by running them on an untouched checkout at `61587c86d`: `linear-cli.test.mjs`
"replica never consulted" and `heartbeat-event.test.mjs` "minus .local" (the latter is
hostname-dependent). Likely the red `execution-core-unit-tests` on main noted in COORD-253.

⚠️ The diff is deliberately kept to **201 insertions / 2 deletions**: `trunk fmt` wanted to rewrite
608 lines of `daemon.mjs` for a 55-line change, so the pre-existing files keep their formatting and
only the new files are formatted.

## Scope

This is the **bridge** the ticket asks for, not the durable fix. The durable one is the lease plane
(CTL-1785 / CTL-1786) — a claim that goes through an authority that can refuse.

Refs CTL-2068, CTC-787, CTC-776, CTL-758, CTL-1174

### CODE REVIEW

**Comment by ryanrozich** [comment]
**Correction to the Verification section, before anyone acts on it.**

I wrote that the two local failures were "likely the red `execution-core-unit-tests` on main noted in COORD-253." **That is wrong, and the check that settles it is cheap:** both files *are* in the workflow's stable list, and `execution-core-unit-tests` on `318268fb9` (main) is **`success`**.

So they are pre-existing **on my machine**, not on main — environment-dependent, not a red trunk:

- `heartbeat-event.test.mjs` "host.name defaults to `os.hostname()` minus `.local`" — hostname-dependent; this host is `office-desk.rozich`.
- `linear-cli.test.mjs` "mode off → linearis, replica never consulted" — this host has a real replica DB at the default path; a CI runner does not.

The substantive claim they supported is unchanged and was verified the right way: **neither is caused by this PR** (both fail identically on an untouched checkout at `61587c86d`). What was wrong was the inference about *main's* state, which nobody should chase.

File: plugins/dev/scripts/execution-core/replica-read.mjs [hunk 1]
```diff
@@ -943,9 +950,26 @@ export function createReplicaReader({ dbPath = getReplicaDbPath() } = {}) {
...
+  // lastStateChange(identifier) → { actorId, toState, atMs } | undefined
+  //   HIT  → the newest row whose to_state is set
+  //   no row / no db / any throw → undefined (fail-open; the CTL-2068 guard reads an
+  //   undefined here as "could not look" and ALLOWS the write, never refuses on it)
+  const lastStateChange = (identifier) => {
+    if (!identifier) return undefined;
+    try {
+      const row = open().prepare(LAST_STATE_CHANGE_SELECT).get(identifier);
+      if (!row) return undefined;
+      return { actorId: row.actorId ?? null, toState: row.toState ?? null, atMs: coerceMs(row.atMs) };
```
[the remark this answers is not in the document]
Comment (ryanrozich) [reply plugins/dev/scripts/execution-core/replica-read.mjs:962]: **P1 accepted and fixed in `9b70b52c7` — and you are more right than the finding says.**

I verified rather than took it on faith, and this repo had already measured the exact numbers: CTL-1847's `linear-feed-diff.mjs:10-17` records `issues.state` at **~11 s** (webhook-fed) against `issue_history` at **~201 s** (reconcile-only, because the mirror's webhook normalizer has no history-synthesis path — CTC-587). The same file records something the finding does not: **140 fleet-wide issues have NO history rows at all** (13 CTL, all 34 POS), so for those the guard could never see an actor no matter how long it waited.

**The fix is your second suggestion:** the row is now trusted only when its `toState` equals the state the ticket is actually in. A row that does not describe the current state is not evidence about who put it there. It is checked **before** the actor branch, so a coincidentally lane-authored older row cannot be read as a live claim either.

⛔ **What I have NOT done, stated plainly rather than implied:** this DECLINES (`inconclusive` → the write proceeds); it does not close the window. Your first suggestion — actor evidence that arrives with the current state — is the real fix and I could not build it here, because **no timely actor source exists**. I checked the obvious candidate before concluding that: `issues.bot_actor_name`/`bot_actor_type` are webhook-fed and therefore timely, but they are **empty for 4,478 of 4,481 non-removed issues** (the only 3 populated rows say `GitHub`/`integration`). The app-actor's own writes do not populate them.

So today's behaviour is **honest and countable** rather than correct: the guard now reports a named `history-row-does-not-match-current-state` instead of confidently returning `LAST_CHANGE_BY_FLEET` and permitting the regression. That is a real change in kind — the previous answer was actively wrong, and wrong in the permissive direction.

Filed as **[CTL-2070](https://linear.app/coalesce-labs/issue/CTL-2070)** with the measurements, the three candidate sources, the trap in the most promising one (the fleet's own write ledger has to survive its *own* 11 s propagation window, so it needs a supersession test rather than a bare equality check), and an AC that pins the timing at 60 seconds with a negative control.

**Residual coverage, so nobody reads more into this PR than it delivers:** any claim older than the reconcile cadence — the ordinary case of a lane holding a ticket for hours — is covered, as are the 2nd and 3rd regressions on CTC-787 itself (the fleet regressed that ticket three times across ~30 minutes, not once). The first ~200 seconds after a claim is not.

One more thing your finding shook loose: my original test fixtures passed `{ actorId }` with **no `toState`**, while the production replica reader always returns it. So the new branch would have been dead in test and live in production. Every fixture now matches the production row shape, and three further mutations are red (decline removed, comparison inverted, `toState` nulled).
